python/tests/stress/arcticdb/version_store/test_mem_leaks.py

```diff
@@ -146,13 +154,132 @@ def grow_exp(df_to_grow: pd.DataFrame, num_times_xx2: int):
         df_to_grow = pd.concat([df_to_grow, df_prev])
     return df_to_grow
 
+def generate_big_dataframe(rows:int=1000000):
+    print("Generating big dataframe")
+    st = time.time()
+    df = get_sample_dataframe(rows)
+    df = grow_exp(df, 5)
+    print("Generation took :", time.time() - st)
+    return df
+
+#endregion
+
+#region HELPER functions for memray tests
+
+def construct_df_querybuilder_tests(size: int) -> pd.DataFrame:
+    df = get_sample_dataframe(size)
+    date_range = pd.date_range(start="2000-1-1", periods=size, freq="s")
+    df.index = pd.Index(date_range)
+    df.index.name = "timestamp"
+    return df
+
+
+def query_group() -> QueryBuilder:
+    """
+        groupby composite aggregation query for QueryBuilder memory tests.
+        The query basically will do aggregation of half of dataframe
+    """
+    q = QueryBuilder()
+    return (
+        q[q["bool"] == True]
+        .groupby("uint8")
+        .agg({"uint32": "mean", "int32": "sum", "strings": "count", "float64": "sum", "float32": "min", "int16": "max"})
+    )
+
+
+def query_group_bigger_result() -> QueryBuilder:
+    """
+        groupby composite aggregation query for QueryBuilder memory tests.
+        The query basically will do aggregation of whole dataframe
+    """
+    q = QueryBuilder()
+    return (
+        q[q["strings"] != "QASDFGH"]
```

Comment:
The chances of this filtering anything out are very low, is that deliberate?

Reply:
yes, this is deliberate to impose attempt to filter by impossible condition which should still be evaluated. And as we know with strings things are always buggier than with numbers

Reply:
My issue is that the condition is not impossible, just unlikely. Making it actually impossible is fine (e.g. use a different character set than the generated dataframe), but we have optimisations in place for when nothing is filtered out, which could then skew the results if you get unlucky and the dataframe does contain `QASDFGH`

Reply:
Yes. It is very unlikely. Thus making the results 99.99% of the dataframe (note that by default I think we generate str of 10 chars array and this is 7 chars. So perhaps really impossible to achieve unless bug

Anyway for our purposes we do not need exactness at all. Thus actual returned result if it is 99 or 100% of rows really does not matter

Reply:
It does matter. Returning 100% and 99.999% of a dataframe can take totally different code paths

Reply:
Reading the comment above I don't understand the purpose of this. What do we want to achieve? Do we want to

1. Filter nothing at all and get an empty data frame?
2. Filter small part of the data frame?

If 1 is the desired option what Alex suggested i.e. using a different charset seems like a better idea.
If 2 is the desired option we should use a string that we're sure filters small amount of data.

Given that randomness is also involved I think there are some unknowns in the tests. @alexowens90 what is your opinion?

Reply:
Lets accept the way it is. If you want me to add additional query I will no problem, but the way I have constructed this is meant to remain

Reply:
Let's see what @alexowens90 thinks

Reply:
added aditional query also without any filter. Thus we have now 3 queries that have groupby agg., with filter(50%) filter(100%) effectivly and no filter ...

there is additional functional test also that covers the case if there is something fishy with queries

Reply:
@grusev the query needs to be changed to one of:

- Return 100% of the dataframe
- Return <100% of the dataframe

And it has to be consistent. We have optimisations in place so that a filter that doesn't remove anything is much more efficient than a filter that removes even 1 row.

Reply:
only the comment is misleading but all is ok